src/Provider/AbstractProvider.php

```diff
@@ -21,11 +24,22 @@
  */
 abstract class AbstractProvider implements ProviderInterface
 {
-    protected $httpAdapter;
+    /**
+     * @var HttpClient
+     */
+    private $httpClient;
+
+    /**
+     * @var RequestFactory
+     */
+    private $httpMessageFactory;
 
-    public function __construct(HttpAdapterInterface $httpAdapter)
+    /**
+     * @param HttpClient $httpClient
+     */
+    public function __construct(HttpClient $httpClient = null)
```

Comment:
> I do not people will care what PSR-7 implementation that are used by Swap internally.

@Nyholm this is the main reason why we have factories in the first place. Otherwise libs could just rely on a chosen implementation. Discovery is cool, but I think constructor injection is better. Also, there is no need to modify the message factory instance after it's injected, also there is no need to have a getter for it IMO.